File: common/src/test/java/io/github/notenoughupdates/moulconfig/ConfigFieldOrderTest.java
package io.github.notenoughupdates.moulconfig;

import io.github.notenoughupdates.moulconfig.annotations.ConfigOrder;
import io.github.notenoughupdates.moulconfig.annotations.ConfigOverride;
import io.github.notenoughupdates.moulconfig.internal.MCLogger;
import io.github.notenoughupdates.moulconfig.internal.Warnings;
import io.github.notenoughupdates.moulconfig.processor.ConfigProcessorDriver;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.lang.reflect.Field;
import java.lang.reflect.Method;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class ConfigFieldOrderTest {
    public static class SingleClass {
        public Object a = new Object();
        public Object b = new Object();
        public Object c = new Object();
    }

    public static class OrderedSingleClass {
        @ConfigOrder(3) public Object c = new Object();
        @ConfigOrder(1) public Object a = new Object();
        @ConfigOrder(2) public Object b = new Object();
    }

    public static class Parent {
        public Object x = new Object();
        public Object y = new Object();
        public Object z = new Object();
    }

    public static class ChildAppendsField extends Parent {
        public Object w = new Object();
    }

    public static class ChildShadowsWithoutAnnotation extends Parent {
        public Object y = new Object();
    }

    public static class ChildShadowsWithAnnotation extends Parent {
        @ConfigOverride public Object y = new Object();
    }

    public static class ParentWithOrderedField {
        public Object a = new Object();
        @ConfigOrder(5) public Object b = new Object();
        public Object c = new Object();
    }

    public static class ChildInheritsParentOrder extends ParentWithOrderedField {
        @ConfigOverride public Object b = new Object();
    }

    public static class ChildExplicitOverrideOrder extends ParentWithOrderedField {
        @ConfigOverride(overrideOrder = 99) public Object b = new Object();
    }

    public static class GrandParent {
        public Object p = new Object();
        public Object q = new Object();
    }

    public static class MiddleParent extends GrandParent {
        public Object r = new Object();
    }

    public static class GrandChild extends MiddleParent {
        @ConfigOverride public Object q = new Object();
    }

    private final Method getSortedFields;
    private final List<String> capturedWarnings = new ArrayList<>();
    private MCLogger previousLogger;

    public ConfigFieldOrderTest() throws NoSuchMethodException {
        getSortedFields = ConfigProcessorDriver.class.getDeclaredMethod("getSortedFields", Class.class);
        getSortedFields.setAccessible(true);
    }

    @BeforeEach
    void installCapturingLogger() {
        previousLogger = Warnings.logger;
        Warnings.shouldWarn = true;
        Warnings.logger = new MCLogger() {
            @Override public void warn(String text) { capturedWarnings.add(text); }
            @Override public void info(String text) {}
            @Override public void error(String text, Throwable throwable) {}
        };
    }

    @AfterEach
    void restoreLogger() {
        Warnings.logger = previousLogger;
        capturedWarnings.clear();
    }

    @SuppressWarnings("unchecked")
    private List<Field> sortedFields(Class<?> type) {
        try {
            return (List<Field>) getSortedFields.invoke(null, type);
        } catch (ReflectiveOperationException e) {
            throw new RuntimeException(e);
        }
    }

    private List<String> fieldNames(Class<?> type) {
        List<String> names = new ArrayList<>();
        for (Field field : sortedFields(type)) {
            names.add(field.getName());
        }
        return names;
    }

    @Test
    void declarationOrderIsPreservedWithoutAnnotations() {
        Assertions.assertEquals(Arrays.asList("a", "b", "c"), fieldNames(SingleClass.class));
    }

    @Test
    void configOrderSortsFieldsWithinAClass() {
        Assertions.assertEquals(Arrays.asList("a", "b", "c"), fieldNames(OrderedSingleClass.class));
    }

    @Test
    void parentFieldsAppearBeforeChildFields() {
        Assertions.assertEquals(Arrays.asList("x", "y", "z", "w"), fieldNames(ChildAppendsField.class));
    }

    @Test
    void shadowingWithoutConfigOverrideEmitsAWarning() {
        sortedFields(ChildShadowsWithoutAnnotation.class);
        Assertions.assertTrue(capturedWarnings.stream().anyMatch(it -> it.contains("y")));
    }

    @Test
    void configOverrideSuppressesShadowWarning() {
        sortedFields(ChildShadowsWithAnnotation.class);
        Assertions.assertTrue(capturedWarnings.stream().noneMatch(it -> it.contains("y")));
    }

    @Test
    void configOverrideSlotsChildFieldIntoParentPosition() {
        List<Field> fields = sortedFields(ChildShadowsWithAnnotation.class);
        List<String> names = new ArrayList<>();
        for (Field field : fields) names.add(field.getName());
        Assertions.assertEquals(Arrays.asList("x", "y", "z"), names);
        Assertions.assertEquals(ChildShadowsWithAnnotation.class, fields.get(1).getDeclaringClass());
    }

    @Test
    void configOverrideWithoutOverrideOrderInheritsParentConfigOrderValue() {
        List<Field> fields = sortedFields(ChildInheritsParentOrder.class);
        List<String> names = new ArrayList<>();
        Field b = null;
        for (Field field : fields) {
            names.add(field.getName());
            if (field.getName().equals("b")) b = field;
        }
        Assertions.assertEquals(Arrays.asList("a", "c", "b"), names);
        Assertions.assertEquals(ChildInheritsParentOrder.class, b.getDeclaringClass());
    }

    @Test
    void configOverrideWithExplicitOverrideOrderUsesThatValueOverInherited() {
        List<String> names = fieldNames(ChildExplicitOverrideOrder.class);
        Assertions.assertEquals("b", names.get(names.size() - 1));
    }

    @Test
    void overrideWorksCorrectlyThroughMultipleInheritanceLevels() {
        List<Field> fields = sortedFields(GrandChild.class);
        List<String> names = new ArrayList<>();
        Field q = null;
        for (Field field : fields) {
            names.add(field.getName());
            if (field.getName().equals("q")) q = field;
        }
        Assertions.assertEquals(Arrays.asList("p", "q", "r"), names);
        Assertions.assertEquals(GrandChild.class, q.getDeclaringClass());
    }
}

File: common/src/test/java/io/github/notenoughupdates/moulconfig/ConfigFieldOrderTest.kt
package io.github.notenoughupdates.moulconfig

import io.github.notenoughupdates.moulconfig.annotations.ConfigOrder
import io.github.notenoughupdates.moulconfig.annotations.ConfigOverride
import io.github.notenoughupdates.moulconfig.internal.Warnings
import io.github.notenoughupdates.moulconfig.processor.ConfigProcessorDriver
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import java.lang.reflect.Field

/**
 * Tests for field ordering and override deduplication in [ConfigProcessorDriver.getSortedFields].
 *
 * Covers:
 * - Declaration order is preserved without annotations
 * - [ConfigOrder] sorts fields within and across classes
 * - Shadowed fields without [ConfigOverride] emit a warning
 * - [ConfigOverride] suppresses that warning and slots into the parent's position
 * - [ConfigOverride] inherits the parent's [ConfigOrder] value by default
 * - [ConfigOverride.overrideOrder] takes precedence over inherited order when set
 * - Correct behavior through multiple levels of inheritance
 */
class ConfigFieldOrderTest {

    open class SingleClass { val a = Unit; val b = Unit; val c = Unit }
    open class OrderedSingleClass { @ConfigOrder(3) val c = Unit; @ConfigOrder(1) val a = Unit; @ConfigOrder(2) val b = Unit }

    open class Parent { val x = Unit; open val y = Unit; val z = Unit }
    open class ChildAppendsField : Parent() { val w = Unit }
    open class ChildShadowsWithoutAnnotation : Parent() { override val y = Unit }
    open class ChildShadowsWithAnnotation : Parent() { @ConfigOverride override val y = Unit }

    open class ParentWithOrderedField { val a = Unit; @ConfigOrder(5) open val b = Unit; val c = Unit }
    open class ChildInheritsParentOrder : ParentWithOrderedField() { @ConfigOverride override val b = Unit }
    open class ChildExplicitOverrideOrder : ParentWithOrderedField() { @ConfigOverride(overrideOrder = 99) override val b = Unit }

    open class GrandParent { val p = Unit; open val q = Unit }
    open class MiddleParent : GrandParent() { val r = Unit }
    open class GrandChild : MiddleParent() { @ConfigOverride override val q = Unit }

    private val getSortedFields = ConfigProcessorDriver::class.java
        .getDeclaredMethod("getSortedFields", Class::class.java)
        .also { it.isAccessible = true }

    private var previousShouldWarn = false
    private var previousShouldCrash = false

    @BeforeEach
    fun captureWarningsState() {
        previousShouldWarn = Warnings.shouldWarn
        previousShouldCrash = Warnings.shouldCrash
    }

    @AfterEach
    fun restoreWarningsState() {
        Warnings.shouldWarn = previousShouldWarn
        Warnings.shouldCrash = previousShouldCrash
    }

    @Suppress("UNCHECKED_CAST")
    private fun sortedFields(type: Class<*>) = getSortedFields.invoke(null, type) as List<Field>
    private fun fieldNames(type: Class<*>) = sortedFields(type).map { it.name }

    private fun assertWarningEmitted(type: Class<*>) {
        Warnings.shouldWarn = true
        Warnings.shouldCrash = true
        assertThrows(RuntimeException::class.java) { sortedFields(type) }
    }

    private fun assertNoWarningEmitted(type: Class<*>) {
        Warnings.shouldWarn = true
        Warnings.shouldCrash = true
        assertDoesNotThrow { sortedFields(type) }
    }

    /** Declaration order of fields within a single class should be preserved when no ordering annotations are present. */
    @Test fun `declaration order is preserved without annotations`() =
        assertEquals(listOf("a", "b", "c"), fieldNames(SingleClass::class.java))

    /** [ConfigOrder] should sort fields by ascending value, regardless of declaration order. */
    @Test fun `ConfigOrder sorts fields within a class`() =
        assertEquals(listOf("a", "b", "c"), fieldNames(OrderedSingleClass::class.java))

    /** Fields declared in a parent class should appear before fields declared in the child. */
    @Test fun `parent fields appear before child fields`() =
        assertEquals(listOf("x", "y", "z", "w"), fieldNames(ChildAppendsField::class.java))

    /** Shadowing a parent field without [ConfigOverride] should emit a warning. */
    @Test fun `shadowing without ConfigOverride emits a warning`() =
        assertWarningEmitted(ChildShadowsWithoutAnnotation::class.java)

    /** [ConfigOverride] should suppress the shadow warning. */
    @Test fun `ConfigOverride suppresses shadow warning`() =
        assertNoWarningEmitted(ChildShadowsWithAnnotation::class.java)

    /** [ConfigOverride] should slot the child field back into the parent field's original position. */
    @Test fun `ConfigOverride slots child field into parent position`() {
        val fields = sortedFields(ChildShadowsWithAnnotation::class.java)
        assertEquals(listOf("x", "y", "z"), fields.map { it.name })
        assertEquals(ChildShadowsWithAnnotation::class.java, fields[1].declaringClass)
    }

    /** [ConfigOverride] without an explicit [ConfigOverride.overrideOrder] should inherit the parent's [ConfigOrder] value. */
    @Test fun `ConfigOverride without overrideOrder inherits parent ConfigOrder value`() {
        val fields = sortedFields(ChildInheritsParentOrder::class.java)
        assertEquals(listOf("a", "b", "c"), fields.map { it.name })
        assertEquals(ChildInheritsParentOrder::class.java, fields.first { it.name == "b" }.declaringClass)
    }

    /** An explicit [ConfigOverride.overrideOrder] value should take precedence over the inherited parent order. */
    @Test fun `ConfigOverride with explicit overrideOrder uses that value over inherited`() =
        assertEquals("b", fieldNames(ChildExplicitOverrideOrder::class.java).last())

    /** [ConfigOverride] should correctly resolve position through multiple levels of inheritance. */
    @Test fun `override works correctly through multiple inheritance levels`() {
        val fields = sortedFields(GrandChild::class.java)
        assertEquals(listOf("p", "q", "r"), fields.map { it.name })
        assertEquals(GrandChild::class.java, fields.first { it.name == "q" }.declaringClass)
    }
}
